File: 0046-permutations/0046-permutations.java
class Solution {
    List<List<Integer>> list;
    public void helper(int idx,int nums[] )
    {
        if(idx>=nums.length)
        {
            List<Integer> temp=new ArrayList<>();
            for(int x:nums)
            temp.add(x);
            list.add(new ArrayList<>(temp));
            return ;
        }
        for(int i=idx;i<nums.length;i++)
        {
            swap(nums,i,idx);
            helper(idx+1,nums);
            swap(nums,i,idx);
        }
    }
    void swap(int nums[],int i,int j)
    {
        int temp=nums[i];
        nums[i]=nums[j];
        nums[j]=temp;
    }
    public List<List<Integer>> permute(int[] nums) {
        list=new ArrayList<>();
        helper(0,nums);
        return list;
    }
}

File: 0046-permutations/0046-permutations.py
class Solution:
    def _init_(self):
        self.lst = []

    def helper(self, idx, nums):
        if idx >= len(nums):
            self.lst.append(nums[:])  # Copy the list and append it to the result
            return

        for i in range(idx, len(nums)):
            nums[idx], nums[i] = nums[i], nums[idx]  # Swap elements
            self.helper(idx + 1, nums)
            nums[idx], nums[i] = nums[i], nums[idx]  # Undo the swap

    def permute(self, nums):
        self.lst = []
        self.helper(0, nums)
        return self.lst